### DISCUSSION TITLE
"Diplomat" Language server and VSCode extension

### DISCUSSION THREAD

**Discussion by suzizecat** [MikePopoloski/slang#1072] [Show and tell]
Hi everyone !

TL;DR : A [language server](https://github.com/suzizecat/slang-lsp-tools) and its [VSCode Extension](https://github.com/suzizecat/diplomat-vscode)

As I am alone on this and have not much known users, the documentation is _minimal_ to say the least and those may have some rough edges but also some interesting features ! 

New users and peoples who actually know how to work with TypeScript and C++ and would take a look at this work would be very welcome !

---

As I have been disappointed with the available VSCode extensions made for working on SystemVerilog (often only supporting Verilog, and almost never supporting the jump across files implied by a "Go To references"), I decided some time ago to try and develop a Language Server based upon Slang that would fulfill my dreams of mass renames and jumping around...

Hence, I developed "Diplomat" (originally a side project to help on another extension: TerosHDL).
This LS (language server) is able to perform:
 - Go to definition/references (across files)
 - Syntax highlighting 
 - Rename symbol
 - Syntax check and lint using Slang capabilities
 - Semi-auto-detection of the design hierarchy (in absence of a designed top-level, all suitable files are passed to slang which sort this mess by itself)
 - Ignore files/path/path patterns
 - Open waveform files with GTKwave and have a kind of synchronization between GTKWave and the editor (this is a proof of concept and will evolve to something actually useful)
 - Some kind of formatting (this is tedious to do)
 - And maybe some features that I have forgotten.

This project is split in two elements:
 - The LS itself, written in modern C++ and leveraging slang: https://github.com/suzizecat/slang-lsp-tools (this also includes some "playground level" stuff and tools to experiment with slang API)
 - The VSCode extension, written in TypeScript ~(please god help me)~ and handling all the VSCode part and UI integration: https://github.com/suzizecat/diplomat-vscode . 

As of today, I am using it as a professional work tool as a digital designer, and it starts to be comfortably usable. 

While it may fail to properly analyze a big workspace simply on opening (I am talking hundreds of files, including UVM verification testbenchs that I have not tried to support yet), setting the top level file and excluding some unused folders usually get the workspace in a working state.

**Comment by suzizecat** [comment]
For reference, while it is a bit rough around the edges, the latest commits allows to see the value of a simulation opened by GTKWave directly within the source code (the values are updated when you change the cursor position in GTKWave):

![image](https://github.com/user-attachments/assets/54a48a58-f105-4dd0-9867-ebd8eabb580e)

**Comment by suzizecat** [comment]
I made a documentation (ongoing) and have setup a release flow on GitHub for both of those projects.
You can find:
 - [the documentation](https://suzizecat.github.io/diplomat-lsp/)
 - [the server releases](https://github.com/suzizecat/slang-lsp-tools/releases)
 - [the client releases](https://github.com/suzizecat/diplomat-vscode/releases)

The server is built on `ubuntu-latest` which is, as the time of writing, 22.04 .
